pkg/operator/controller/ingress/controller.go

```diff
@@ -713,6 +726,121 @@ func (r *reconciler) ensureIngressDeleted(ingress *operatorv1.IngressController)
 	return utilerrors.NewAggregate(errs)
 }
 
+func (r *reconciler) clearRouteStatusForDeletedIngress(icName string) error {
+	// Clear route status for routes that have been admitted by this ingress controller.
+	// List all routes.
+	start := time.Now()
+	routeList := &routev1.RouteList{}
+	routesCleared := 0
+	if err := r.cache.List(context.TODO(), routeList); err != nil {
+		return fmt.Errorf("failed to list all routes in order to clear route status")
+	}
+	for _, route := range routeList.Items {
+		cleared, err := r.clearRouteStatus(route, icName)
+		if cleared {
+			routesCleared++
+		}
+		if err != nil {
+			return err
+		}
+	}
+	elapsed := time.Since(start)
+	log.Info("cleared route status for ingress deletion", "Ingress Controller", icName, "Routes Status Cleared", routesCleared, "Time Elapsed", elapsed)
+	return nil
+}
+
+// Clears this route's status that is admitted by provided ingressController
+func (r *reconciler) clearRouteStatus(route routev1.Route, icName string) (bool, error) {
+	// Go through each route and clear status if admitted by this ingress controller.
+	var updated routev1.Route
+	clearedStatus := false
+	for i := range route.Status.Ingress {
+		if condition := findCondition(&route.Status.Ingress[i], routev1.RouteAdmitted); condition != nil {
+			if routerName := &route.Status.Ingress[i].RouterName; routerName != nil && *routerName == icName {
+				// Remove this status since it matches our routerName.
+				route.DeepCopyInto(&updated)
+				updated.Status.Ingress = append(route.Status.Ingress[:i], route.Status.Ingress[i+1:]...)
+				if err := r.client.Status().Update(context.TODO(), &updated); err != nil {
+					return clearedStatus, fmt.Errorf("failed to clear route status for %s with routerName %s: %v", route.Name, icName, err)
+				}
+				clearedStatus = true
+			}
+		}
+	}
+
+	return clearedStatus, nil
+}
+
+// Clear routes that are not owned by this ingress controller
+func (r *reconciler) clearRoutesNotAdmittedByIngress(ingress *operatorv1.IngressController) {
+	// TODO: THIS IS TESTING CODE. NEEDS A LOT OF WORK
+	start := time.Now()
+
+	routesCleared := 0
+	// List all routes.
+	routeList := &routev1.RouteList{}
+	if err := r.cache.List(context.TODO(), routeList); err != nil {
+		fmt.Errorf("failed to list all routes in order to clear route status")
+	}
+	for _, route := range routeList.Items {
+		routeSelectorMatch := false
+		namespaceSelectorMatch := false
+
+		// TODO: REAL MESSY, FIX THIS UP
+		routeSelector, _ := metav1.LabelSelectorAsSelector(ingress.Spec.RouteSelector)
+		if routeSelector != nil {
+			routeLabelSelector, _ := labels.Parse(routeSelector.String())
+			if routeLabelSelector == nil {
+				routeSelectorMatch = true
+			} else if routeLabelSelector.Matches(labels.Set(route.Labels)) {
+				routeSelectorMatch = true
+			}
+		} else {
+			routeSelectorMatch = true
+		}
```

Comment:
Apply the selector when you list routes:
```suggestion
	routeSelector, err := metav1.LabelSelectorAsSelector(ingress.Spec.RouteSelector)
	if err != nil {
		return fmt.Errorf("ingresscontroller %s/%s has an invalid route selector: %w", ingress.Namespace, ingress.Name, err)
	}
	listOpts := []client.ListOption{
		client.MatchingLabelsSelector{Selector: routeSelector},
	}
	if err := r.cache.List(context.TODO(), routeList, listOpts); err != nil {
		fmt.Errorf("failed to list all routes in order to clear route status")
	}
	for _, route := range routeList.Items {
```

Reply:
Sorry, my suggestion isn't right; you would need to list routes using the _old_ route selector, if you had it.  :thinking:...

Reply:
I don't think it's possible to negate a selector in the general case because a selector is conjunction, and the negation of a conjunction is a disjunction, which cannot be represented as a selector.  So what you could do is list routes using the route selector to get a filtered list (as I described in my earlier comment), then list _all_ routes, and then iterate through the list of all routes and clear the status of any that aren't in the filtered list.  This would require a little bit of cleverness to do efficiently though; maybe something like the following.  Build a set from the filtered list:
```go
	type routeKey struct {
		name, namespace string
	}
	routesInShard := map[routeKey]struct{}{}
	for i := range filteredRouteList.Items {
		route := &filteredRouteList.Items[i]
		routesInShard[routeKey{route.Name, route.Namespace}] = struct{}{}
	}
```
Then use this set when iterating over the full list of routes:
```go
	for i := range routeList.Items {
		route := &routeList.Items[i]
		if _, ok := routesInShard[routeKey{route.Name, route.Namespace]; !ok {
			r.clearRouteStatus(...)
```

Reply:
Done